addons/account/models/account_invoice.py

```diff
@@ -164,9 +164,27 @@ def _get_outstanding_info_JSON(self):
                     # get the outstanding residual value in invoice currency
                     if line.currency_id and line.currency_id == self.currency_id:
                         amount_to_show = abs(line.amount_residual_currency)
-                    else:
+                    elif line.currency_id and line.currency_id != self.currency_id:
+                        amount_to_show = line.currency_id._convert(
+                            abs(line.amount_residual_currency),
+                            self.currency_id,
+                            self.company_id,
+                            self.move_id.date or fields.Date.today()
```

Comment:
There is only 2 cases:

All in foreign currency => use amount_residual_currency.
Otherwise convert from company's currency to invoice's currency using the payment's date.
```
if self.currency_id != self.company_id.currency_id and line.currency_id == self.currency_id:
  amount_to_show = abs(line.amount_residual_currency)
else:
  amount_to_show = self.company_id.currency_id._convert(
    abs(line.amount_residual),
    self.currency_id,
    self.company_id,
    line.date
  )
```